link.go

```diff
@@ -290,13 +271,39 @@ func (l *link) attach(ctx context.Context, s *Session) error {
 
 	// send Attach frame
 	log.Debug(1, "TX (attachLink): %s", attach)
-	_ = s.txFrame(attach, nil)
+
+	// we use send to have positive confirmation on transmission
+	send := make(chan encoding.DeliveryState)
+	_ = s.txFrame(attach, send)
 
 	// wait for response
 	var fr frames.FrameBody
 	select {
 	case <-ctx.Done():
-		// TODO: this leaks l's handle
+		select {
+		case <-send:
+			// attach was written to the network. assume it was received
+			// and that the ctx was too short to wait for the ack. in this
+			// case we must send a detach before deallocation
+			go func() {
+				_ = s.txFrame(&frames.PerformDetach{
+					Handle: l.Handle,
+					Closed: true,
+				}, nil)
+				select {
+				case <-s.done:
+					// session has terminated
+				case <-time.After(5 * time.Second):
+					// timed out
+				case <-l.RX:
+					// received ack
+				}
+				s.deallocateHandle(l)
+			}()
+		default:
```

Comment:
If we timed out and we get the DETACH frame after we've deallocated the handle what happens? Does it cause any issues or will it just get ignored?